pkg/vulnsrc/openeuler/openeuler.go

```diff
@@ -0,0 +1,246 @@
+
+	return nil
+}
+
+func (vs VulnSrc) save(cvrfs []Cvrf) error {
+	err := vs.dbc.BatchUpdate(func(tx *bolt.Tx) error {
+		return vs.commit(tx, cvrfs)
+	})
+	if err != nil {
+		return xerrors.Errorf("error in batch update: %w", err)
+	}
+	return nil
+}
+
+func (vs VulnSrc) commit(tx *bolt.Tx, cvrfs []Cvrf) error {
+	for _, cvrf := range cvrfs {
+		affectedPkgs := getAffectedPackages(cvrf.ProductTree)
+		if len(affectedPkgs) == 0 {
+			continue
+		}
+
+		for _, affectedPkg := range affectedPkgs {
+			advisory := types.Advisory{
+				FixedVersion: affectedPkg.Package.FixedVersion,
+			}
+
+			if err := vs.dbc.PutDataSource(tx, affectedPkg.OSVer, source); err != nil {
+				return xerrors.Errorf("failed to put data source: %w", err)
+			}
+			if err := vs.dbc.PutAdvisoryDetail(tx, cvrf.Tracking.ID, affectedPkg.Package.Name,
+				[]string{affectedPkg.OSVer}, advisory); err != nil {
+				return xerrors.Errorf("unable to save %s CVRF: %w", affectedPkg.OSVer, err)
+			}
+		}
+
+		var references []string
+		for _, ref := range cvrf.References {
+			references = append(references, ref.URL)
+		}
+
+		severity := types.SeverityUnknown
+		for _, cvuln := range cvrf.Vulnerabilities {
+			for _, threat := range cvuln.Threats {
+				sev := severityFromThreat(threat.Severity)
+				if severity < sev {
+					severity = sev
+				}
+			}
+		}
+
+		vuln := types.VulnerabilityDetail{
+			References:  references,
+			Title:       cvrf.Title,
+			Description: getDetail(cvrf.Notes),
+			Severity:    severity,
+		}
+		if err := vs.dbc.PutVulnerabilityDetail(tx, cvrf.Tracking.ID, source.ID, vuln); err != nil {
+			return xerrors.Errorf("failed to save openEuler CVRF vulnerability: %w", err)
+		}
+
+		// for optimization
+		if err := vs.dbc.PutVulnerabilityID(tx, cvrf.Tracking.ID); err != nil {
+			return xerrors.Errorf("failed to save the vulnerability ID: %w", err)
+		}
+	}
+	return nil
+}
+
+func getAffectedPackages(productTree ProductTree) []AffectedPackage {
+	var pkgs []AffectedPackage
+
+	for _, branch := range productTree.Branches {
+		if branch.Type != "Package Arch" || branch.Name != "aarch64" {
+			continue
+		}
+		for _, production := range branch.Productions {
+			osVer := getOSVersion(production.CPE)
+			if osVer == "" {
+				continue
+			}
+
+			productID := production.ProductID
+			if productID == "" {
+				parts := strings.Split(production.Text, ".oe")
+				productID = parts[0]
+			}
+
+			pkg := getPackage(productID)
+			if pkg == nil {
+				log.Printf("invalid package name")
+				continue
+			}
+
+			pkgs = append(pkgs, AffectedPackage{
+				OSVer:   osVer,
+				Package: *pkg,
+			})
+		}
+	}
+	return pkgs
+}
+
+func getOSVersion(cpe string) string {
+	// e.g. cpe:/a:openEuler:openEuler:22.03-LTS-SP3
+	parts := strings.Split(cpe, ":")
+	if len(parts) != 5 || parts[2] != "openEuler" {
+		return ""
+	}
+	version := parts[4]
+	substrings := strings.Split(version, "-")
+	// e.g. 23.09, 22.03-LTS, 22.03-LTS-SP3
+	if len(substrings) < 1 || len(substrings) > 3 {
```

Comment:
len(substrings) is always one or more:
https://go.dev/play/p/Vj_EhC1fH6f

```suggestion
	// e.g. 23.09, 22.03-LTS, 22.03-LTS-SP3
	if len(strings.Split(version, "-")) > 3 {
```